src/pubsub/ua_pubsub_securitygroup.c

```
@@ -0,0 +1,322 @@
+        if(UA_String_equal(&securityGroupName, &tmpSG->config.securityGroupName))
+            return tmpSG;
+    }
+    return NULL;
+}
+
+UA_StatusCode
+UA_SecurityGroupConfig_copy(const UA_SecurityGroupConfig *src,
+                            UA_SecurityGroupConfig *dst) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    memcpy(dst, src, sizeof(UA_SecurityGroupConfig));
+    if(UA_String_copy(&src->securityGroupName, &dst->securityGroupName) !=
+       UA_STATUSCODE_GOOD)
+        return UA_STATUSCODE_BAD;
+
+    if(UA_String_copy(&src->securityPolicyUri, &dst->securityPolicyUri) !=
+       UA_STATUSCODE_GOOD)
+        return UA_STATUSCODE_BAD;
+    return retval;
+}
+
+static UA_StatusCode
+generateKeyData(const UA_PubSubSecurityPolicy *policy, UA_ByteString *key) {
+    if(!key || !policy)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    UA_StatusCode retVal;
+
+    /* Can't not found in specification for pubsub key generation, so use the idea of
+     * securechannel, see specification 1.0.3 6.7.5 Deriving keys for more details
+     In pubsub we do get have OpenSecureChannel request, so we cannot have Client or Server Nonce*/
+    UA_Byte secretBytes[UA_PUBSUB_KEYMATERIAL_NONCELENGTH];
+    UA_ByteString secret;
+    secret.length = UA_PUBSUB_KEYMATERIAL_NONCELENGTH;
+    secret.data = secretBytes;
+
+    UA_Byte seedBytes[UA_PUBSUB_KEYMATERIAL_NONCELENGTH];
+    UA_ByteString seed;
+    seed.data = seedBytes;
+    seed.length = UA_PUBSUB_KEYMATERIAL_NONCELENGTH;
+    memset(seed.data, 0, seed.length);
+    retVal = policy->symmetricModule.generateNonce(policy->policyContext, &secret);
+    retVal |= policy->symmetricModule.generateNonce(policy->policyContext, &seed);
+    if(retVal != UA_STATUSCODE_GOOD)
+        return retVal;
+
+    retVal = policy->symmetricModule.generateKey(policy->policyContext, &secret, &seed, key);
+    return retVal;
+}
+
+static void
+updateSKSKeyStorage(UA_Server *server, UA_SecurityGroup *securityGroup){
```

Silent failures shouldn't happen.
If you cannot return a status code here, then use the logging from the server to warn users.
The goto error is only used once. That should be move inline to reduce the code flow complexity.